README.txt
qrencoder
    by Jacob Harris
    http://www.nimblecode.com/
    http://nycrb.rubyforge.org/qrencoder

== DESCRIPTION:

This Gem is a wrapper around an useful open-source library for creating QR
Codes, a two-dimensional bar code format popular in Japan (and readable by cell
phones even) created by the Denso-Wave Corporation in 1994. These bar codes look
like the following:

http://www.denso-wave.com/qrcode/images/qrcode.gif

The specification for QR codes is readable at
http://www.denso-wave.com/qrcode/index-e.html. QR Code is not the only 2-D
barcode standard in existence, and it is in wider use in Japan than elsewhere.
Notable competitors include PDF417, Semacode/DataMatrix, and Maxi Code (seen on
UPS labels) in North America; ShotCode in the UK; and an additional format used
in Korea. All vary in look and capacity, but QR code has one of the largest
information densities and capacities among them with the following maximum data

* Numeric	7,089
* Alphanumeric 4,296
* 8-Bit 2,953
* Kanji 1,817

All of these in a square that may range from 21 to 177 pixels a side (there are
40 different "versions" of the code that specify different sizes). In addition,
multiple levels of error correction are possible which might also influence the
final size.

== FEATURES/PROBLEMS:
  
* This gem requires you to build and install an external C library
* The QREncode lib this GEM is built around is NOT thread-safe!

== SYNOPSIS:

There are 4 initialization methods for creating a QR code, two for strings, two
for data. Both of these come in a simpler variant which take the string/data to
be encoded. The second argument is a QR Code version (used to pick the size of
the final QR Code); this version is adjusted upwards if the data is too large to
fit within the specified version.

  img = QRCode.encode_string(text, 1)
  puts "#{img.version}"  #=> 7
  puts "#{img.width}"    #=> 34
  img.save_png("/tmp/foo.png")

You can retrieve the pixels of the output image directly or use the included
methods to save a PNG file directly.

== REQUIREMENTS:

This gem requires you to build the C library lib <tt>libqrcode.a</tt> available
from http://megaui.net/fukuchi/works/qrencode/index.en.html. 

Normally, the build process also expects you to install <tt>libpng</tt>, but
this is only used for the <tt>qrenc</tt> command-line utility and can be avoided
if you run the following build sequence:

  ./configure --without-tools
  make
  sudo make install

This gem also requires the following gems to be installed:

* ruby-inline
* png

== INSTALL:

* First build the <tt>libqrencode</tt> library following the instructions above.
* Then <tt>sudo gem install qrencoder --include-dependencies</tt>

== QRCODE LICENSE:

The QR Code specification was created and patented by the Denso Corporation of
Japan. Although the Denso Corporation retains their patent, it is "open in the
sense that the specification of QR Code is disclosed and that the patent right
owned by Denso Wave is not exercised." according to their website.

== LICENSE:

(The MIT License)

Copyright (c) 2007

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.4.1 / 2012-02-02

* Updated for libqrencode 3.2.0

== 1.4.0 / 2011-10-02

* API changes
  * PNG is now a full-fledged object to allow further
    customization. See chunky_png for more details.
  * Update chunky_png dependency to allow newest versions

* Bugfixes
  * alphanumeric mode was not working correctly
  * numeric mode segfaulted when passed an integer
  * Passing an invalid string now raises an ArgumentError
    rather than segfaulting

== 1.3.4 / 2011-03-08

* 1 Minor enhancement
  * Add pixels_per_module option for enlarging the PNG image by the
    provided factor.

== 1.3.3 / 2011-03-08

* Fix bug with C extension location

== 1.3.2 / 2011-03-08

* Bugfix release. Gem erroneously included pre-compiled file.

== 1.3.1 / 2011-03-08

* Bugfix release

== 1.3.0 / 2011-03-08

* 1 Minor enhancement
  * Use more maintain ChunkyPNG library instead of PNG

== 1.2.0 / 2011-02-23

* Major API change
  * Everything is now wrapped in the QREncoder module
  * QRCode.encode_string and QRCode.encode_string_ex are now handled
    through a more conventional object initializer. See QRCode#initialize
    for documentation.
  * QREncoder includes a convenience method for creating QRCode objects.
    See QREncoder.encode for details.

== 1.0.1 / 2008-03-11

* 1 Minor enhancement
  * Updated to using Hoe's new rakefile format

== 1.0.0 / 2007-01-03

* Initial release
  * Wooo!


spec/fixtures.rb
module QREncoder
  module Fixtures
    String1 = [193, 193, 193, 193, 193, 193, 193, 192, 133, 2, 2, 2, 3, 192, 193, 193, 193, 193, 193, 193, 193, 193, 192, 192, 192, 192, 192, 193, 192, 133, 3, 2, 3, 2, 192, 193, 192, 192, 192, 192, 192, 193, 193, 192, 193, 193, 193, 192, 193, 192, 133, 3, 3, 3, 3, 192, 193, 192, 193, 193, 193, 192, 193, 193, 192, 193, 193, 193, 192, 193, 192, 133, 2, 3, 2, 3, 192, 193, 192, 193, 193, 193, 192, 193, 193, 192, 193, 193, 193, 192, 193, 192, 132, 3, 3, 3, 3, 192, 193, 192, 193, 193, 193, 192, 193, 193, 192, 192, 192, 192, 192, 193, 192, 133, 2, 3, 2, 3, 192, 193, 192, 192, 192, 192, 192, 193, 193, 193, 193, 193, 193, 193, 193, 192, 145, 144, 145, 144, 145, 192, 193, 193, 193, 193, 193, 193, 193, 192, 192, 192, 192, 192, 192, 192, 192, 132, 2, 2, 2, 2, 192, 192, 192, 192, 192, 192, 192, 192, 133, 133, 132, 132, 133, 133, 145, 132, 132, 2, 2, 2, 3, 132, 132, 133, 132, 133, 133, 133, 133, 2, 3, 3, 2, 2, 3, 144, 3, 3, 3, 3, 2, 3, 3, 3, 2, 2, 3, 3, 2, 2, 3, 2, 2, 3, 3, 2, 145, 3, 2, 2, 3, 2, 3, 3, 2, 2, 2, 3, 2, 3, 2, 2, 3, 3, 3, 2, 2, 144, 2, 2, 2, 3, 3, 3, 2, 2, 3, 2, 2, 2, 3, 3, 3, 2, 3, 3, 2, 2, 145, 2, 3, 3, 2, 3, 2, 2, 3, 3, 3, 3, 2, 3, 3, 192, 192, 192, 192, 192, 192, 192, 192, 129, 3, 3, 3, 3, 2, 2, 2, 2, 2, 3, 2, 3, 193, 193, 193, 193, 193, 193, 193, 192, 132, 2, 2, 3, 2, 2, 3, 3, 3, 3, 3, 3, 2, 193, 192, 192, 192, 192, 192, 193, 192, 133, 3, 3, 2, 2, 3, 3, 2, 2, 2, 2, 2, 3, 193, 192, 193, 193, 193, 192, 193, 192, 133, 2, 3, 2, 3, 3, 3, 3, 2, 2, 2, 2, 3, 193, 192, 193, 193, 193, 192, 193, 192, 132, 3, 2, 2, 3, 3, 3, 2, 3, 2, 2, 3, 3, 193, 192, 193, 193, 193, 192, 193, 192, 132, 3, 2, 2, 3, 3, 2, 2, 3, 2, 2, 2, 2, 193, 192, 192, 192, 192, 192, 193, 192, 133, 2, 3, 3, 3, 2, 2, 3, 2, 2, 2, 2, 2, 193, 193, 193, 193, 193, 193, 193, 192, 133, 3, 3, 3, 2, 2, 3, 3, 3, 2, 2, 2, 3]
  end
end


spec/qrcode_spec.rb
require 'spec_helper'
require 'fixtures'

describe QREncoder::QRCode do

  describe "#data" do
    context "with default options and 'String'" do
      subject { QREncoder.encode("String").data }
      it { should be_kind_of(Array) }
      it "generates the correct data array" do
        subject.should == QREncoder::Fixtures::String1
      end
    end
    context "with version 4 and 'String'" do
      subject { QREncoder.encode("String", :version => 4).data }
      it { should be_kind_of(Array) }
      it "generates the correct data array" do
        subject.should == QREncoder::Fixtures::String4
      end
    end
  end

  describe "#dup" do
    subject { QREncoder.encode("something").dup }
    its(:width) { should == 21 } 
  end

  describe "#width" do
    subject { QREncoder.encode("something").width }
    it { should == 21 } 
  end

  describe "#version" do

    context "when not specified" do
      subject { QREncoder.encode(string).version }
      context "and string fits within version 1" do
        let(:string) { "hello" }
        it { should == 1 }
      end
      context "and string is too large for version 1" do
        let(:string) { "hello" }
        it { should == 1 }
      end
    end

    context "when specified" do
      subject { QREncoder.encode(string, :version => 2).version }
      context "and string fits within specified version" do
        let(:string) { "hello" }
        it { should == 2 }
      end
      context "and string is too large for specified version" do
        let(:string) { "Lorem ipsum dolor et al and then some more content" }
        it { should == 3 }
      end
    end

  end

  describe "#width" do
    subject { QREncoder.encode("test", :version => version).width }
    context "with version 1" do
      let(:version) { 1 }
      it { should == 21 }
    end
    context "with version 2" do
      let(:version) { 2 }
      it { should == 25 }
    end
    context "with version 3" do
      let(:version) { 3 }
      it { should == 29 }
    end
    context "with version 4" do
      let(:version) { 4 }
      it { should == 33 }
    end
    context "with version 5" do
      let(:version) { 5 }
      it { should == 37 }
    end
    context "with version 6" do
      let(:version) { 6 }
      it { should == 41 }
    end
    context "with version 11" do
      let(:version) { 11 }
      it { should == 61 }
    end
    context "with version 15" do
      let(:version) { 15 }
      it { should == 77 }
    end
    context "with version 40 (the maximum)" do
      let(:version) { 40 }
      it { should == 177 }
    end
  end

  describe "#height" do
    subject { QREncoder.encode("test") }
    it "should be the same as the width" do
      subject.height.should == subject.width
    end
  end

  describe "#pixels" do
    let(:qrcode) { QREncoder.encode("hi") }
    let(:pixels) do
      arr = []
      qrcode.data.each_slice(qrcode.width) do |a|
        arr << a.map { |p| p & 0x1 }
      end
      arr
    end

    it "provides pixel data" do
      qrcode.pixels.should == pixels
    end
  end

  describe "#points" do
    let(:qrcode) { QREncoder.encode("hi") }
    let(:points) do
      arr = []
      y = 0
      qrcode.data.each_slice(qrcode.width) do |r|
        x = 0;
        r.each do |p|
          if (p & 0x1) == 1
            arr << [x, y]
          end
          x += 1
        end
        y += 1
      end
      arr
    end

    it "provides point data" do
      qrcode.points.should == points
    end
  end

  describe "#canvas" do
    let(:qrcode) { QREncoder.encode("hi") }

    it "returns an instance of ChunkyPNG::Canvas" do
      qrcode.canvas.should be_kind_of(ChunkyPNG::Canvas)
    end

    context "with no options specified" do
      subject { qrcode.canvas }
      its(:width) { should == qrcode.width + (4 * 2) }
      specify "background should be white" do
        subject.get_pixel(1,1).should == ChunkyPNG::Color::WHITE
      end
    end

    context "with margin option of 8" do
      subject { qrcode.canvas(:margin => 8) }
      its(:width) { should == qrcode.width + (8 * 2) }
    end

    context "with transparent set to true" do
      subject { qrcode.canvas(:transparent => true) }
      specify "background should be transparent" do
        subject.get_pixel(1,1).should == ChunkyPNG::Color::TRANSPARENT
      end
    end
  end

  describe "#png" do
    let(:qrcode) { QREncoder.encode("hi") }

    it "returns an instance of ChunkyPNG" do
      qrcode.png.should be_kind_of(ChunkyPNG::Image)
    end

    it "sends options to #canvas" do
      canvas = ChunkyPNG::Canvas.new(1,1)
      options = { :margin => 5, :transparent => true }
      qrcode.should_receive(:canvas).with(options).and_return(canvas)
      qrcode.png(options)
    end
  end

  describe "#save_png" do
    let(:qrcode) { QREncoder.encode("hi") }
    let(:path) do
      File.expand_path("../../tmp/test.png", __FILE__)
    end

    before { File.unlink(path) if File.file?(path) }

    it "saves a png image to path specified" do
      qrcode.save_png(path)
      File.open(path) { true }
      File.size(path).should_not be_zero
    end
  end

end


spec/spec_helper.rb
require 'rubygems'
require 'bundler/setup'
require 'rspec'

root = File.expand_path('../..', __FILE__)
Dir["#{root}/spec/support/**/*.rb"].each {|f| require f}

Bundler.require


lib/qrencoder.rb
require 'qrencoder/version'
require 'qrencoder/qrencoder_ext'
require 'qrencoder/qrcode'

module QREncoder
  # Numeric-only encoding mode
  QR_MODE_NUM = 0

  # Alphanumeric-only encoding mode
  QR_MODE_AN = 1

  # 8-bit ASCII encoding mode
  QR_MODE_8 = 2

  # Kanji encoding mode
  QR_MODE_KANJI = 3

  # Low error correction
  QR_ECLEVEL_L = 0

  # Medium error correction
  QR_ECLEVEL_M = 1

  # Medium-high error correction
  QR_ECLEVEL_Q = 2

  # High error correction
  QR_ECLEVEL_H = 3

  # Encode a string, return a QRCode object
  #
  # This method takes 2 arguments: the string you wish to encode, and a hash of
  # options. The options are as follows:
  #
  # [:version] An integer representing the minimum QRCode version (default: +1+)
  # [:correction] The amount of error correction to apply. One of +:low+, +:medium+, +:quarter+, +:high+. (default: +:low+)
  # [:mode] The encoding mode to use. Must be one of +:numeric+, +:alphanumeric+, +:ascii+, +:kanji+. (default: +:ascii+)
  # [:case_sensitive] Set to +false+ if case does not matter. (default: +true+)
  #
  # For more information about what each of these modes and correction levels
  # mean, see QRCode.new
  #
  def self.encode(string, options={})
    version = options[:version] || 1
    correction = corrections[options[:correction] || :low]
    mode = modes[options[:mode] || :ascii]
    case_sensitive = options[:case_sensitive] == false ? 0 : 1
    string.upcase! if mode == QR_MODE_AN
    QRCode.new(string.to_s, version, correction, mode, case_sensitive)
  end

  def self.corrections
    {
      :low => QR_ECLEVEL_L,
      :medium => QR_ECLEVEL_M,
      :quarter => QR_ECLEVEL_Q,
      :high => QR_ECLEVEL_H
    }
  end

  def self.modes
    {
      :numeric => QR_MODE_NUM,
      :alphanumeric => QR_MODE_AN,
      :ascii => QR_MODE_8,
      :kanji => QR_MODE_KANJI
    }
  end

end


lib/qrencoder/png.rb
require 'chunky_png'

module QREncoder
  class PNG
    attr_reader :canvas,
      :background,
      :margin,
      :pixels_per_module,
      :points,
      :width

    ##
    # Takes an optional hash of options. See QRCode#png for details.
    def initialize(qrcode, options={})
      @points = qrcode.points
      @margin = options[:margin] || 4
      @pixels_per_module= options[:pixels_per_module] || 1
      @background = options[:transparent] ? ChunkyPNG::Color::TRANSPARENT : ChunkyPNG::Color::WHITE
      @width = (qrcode.width + (2 * margin)) * pixels_per_module

      @canvas = ChunkyPNG::Image.new(width, width, background)
      plot_points
    end

    def method_missing(method, *args, &block)
      if canvas.respond_to?(method)
        canvas.send(method, *args, &block)
      else
        super
      end
    end

    def respond_to?(method)
      super || canvas.respond_to?(method)
    end

    private
    def plot_points
      points.each do |point|
        pixels_per_module.times do |x_offset|
          x = (point[0] + margin) * pixels_per_module + x_offset
          pixels_per_module.times do |y_offset|
            y = (point[1] + margin) * pixels_per_module + y_offset
            canvas[x,y] = ChunkyPNG::Color::BLACK
          end
        end
      end
    end

  end
end


lib/qrencoder/qrcode.rb
require 'qrencoder/png'

module QREncoder
  # Stores and represents data, points, and/or pixels for a QRCode
  class QRCode
    ##
    # Shortcut to save the QRcode as a PNG at +path+ using default options.
    #
    # For more control, or to specify non-default options, see QRCode#png
    def save_png(path)
      png.save(path)
    end

    ##
    # Returns an instance of PNG, which can be saved to a file with PNG#save or
    # converted to a blob for inline file transfer with PNG#to_blob.
    #
    # Options:
    #
    # [:margin] A pixel value for the margin around each side of the code. This should be 4 or greater. (default: +4+)
    # [:transparent] Background transparency. Can be true or false. (default: +false+)
    # [:pixels_per_module] Adjusts the entire PNG image by the given factor, integer. (default: +1+)
    #
    def png(options={})
      PNG.new(self, options)
    end

    def canvas(options={})
      png(options).canvas
    end
  end
end


ext/qrencoder_ext/qrencoder_ext.c
#include "qrencode.h"
#include "ruby.h"
#include "errno.h"
#include "string.h"

VALUE mQREncoder;
VALUE cQRCode;
int errno;

static void qrcode_free(void *p) {
  QRcode *qrcode = (QRcode *) p;
  QRcode_free(qrcode);
}

/*
 * call-seq:
 *   width
 *
 * Width of the symbol in modules. This value usually corresponds to 1 module
 * is 1 pixel, but you could conceivably scale it up if you wanted to.
 */
static VALUE _width(VALUE self) {
  QRcode *qrcode;
  Data_Get_Struct(self, QRcode, qrcode);
  return INT2FIX(qrcode->width);
}


/*
 * call-seq:
 *   version
 *
 * Version of the symbol. A QR Code version indicates the size of the 2-D
 * barcode in modules. See +qrencode_string+ for a more detailed description of
 * the version. Note that the version returned might be larger than the version
 * specified for an encode_string if the requested version is for a barcode too
 * small to encode the data.
 */
static VALUE _version(VALUE self) {
  QRcode *qrcode;
  Data_Get_Struct(self, QRcode, qrcode);
  return INT2FIX(qrcode->version);
}

/*
 * call-seq:
 *   data
 *
 * Returns the raw data of the QRcode within a single array of width*width
 * elements. Each item is a byte of data of which only the least significant
 * bit is the pixel. The full use of each bit from Least Significant to Most is
 * as follows
 *
 * - 1=black / 0=white
 * - data and ecc code area
 * - format information
 * - version information
 * - timing pattern
 * - alignment pattern
 * - finder pattern and separator
 * - non-data modules (format, timing, etc.)
 *
 * This structure allows the QRcode spec to store multiple types of information
 * within the allocated output buffers, but you usually only care about the pixel
 * color. For those cases, just use the +pixels+ or +points+ methods.
 */
static VALUE _data(VALUE self) {
  QRcode *qrcode;
  VALUE out;
  unsigned char *p, b;
  int i, max;

  Data_Get_Struct(self, QRcode, qrcode);
  p = qrcode->data;
  max = qrcode->width * qrcode->width;
  out = rb_ary_new2(max);

  for (i=0; i < max; i++) {
    b = *p;
    rb_ary_push(out, INT2FIX(b));
    p++;
  }

  return out;
}

/*
 * call-seq:
 *   pixels
 *
 * Returns the QRcode as an array of rows where each item in a row represents
 * the value of the pixel (1 = black, 0 = white)
 */
static VALUE _pixels(VALUE self) {
  QRcode *qrcode;
  VALUE out, row;
  unsigned char *p;
  int x, y, bit;

  Data_Get_Struct(self, QRcode, qrcode);
  p = qrcode->data;
  out = rb_ary_new2(qrcode->width);

  for (y=0; y < qrcode->width; y++) {
    row = rb_ary_new2(qrcode->width);

    for (x=0; x < qrcode->width; x++) {
      bit = *p & 1;
      rb_ary_push(row, INT2FIX(bit));
      p++;
    }

    rb_ary_push(out, row);
  }

  return out;
}

/*
 * call-seq:
 *   points
 *
 * Returns the black pixels of the encoded image as an array of coordinate pairs.
 */
static VALUE _points(VALUE self) {
  QRcode *qrcode;
  VALUE out, point;
  unsigned char *p;
  int x, y, bit;

  Data_Get_Struct(self, QRcode, qrcode);
  p = qrcode->data;
  out = rb_ary_new2(qrcode->width);

  for (y=0; y < qrcode->width; y++) {
    for (x=0; x < qrcode->width; x++) {
      bit = *p & 1;

      if (bit) {
        point = rb_ary_new2(2);
        rb_ary_push(point, INT2FIX(x));
        rb_ary_push(point, INT2FIX(y));
        rb_ary_push(out, point);
      }

      p++;
    }
  }

  return out;
}

/* Handles +clone+ and +dup+ */
static VALUE qr_init_copy(VALUE copy, VALUE orig) {
  QRcode *copy_qrcode;
  QRcode *orig_qrcode;
  if (copy == orig)
    return copy;

  Data_Get_Struct(orig, QRcode, orig_qrcode);
  Data_Get_Struct(copy, QRcode, copy_qrcode);
  MEMCPY(copy_qrcode, orig_qrcode, QRcode, 1);

  /* Code will be freed by original object */
  RDATA(copy)->dfree = NULL;

  return copy;
}

/*
 * call-seq:
 *   new(string, version, eclevel, mode, case_sensitive)
 *
 * Encodes a QR code from a string.
 *
 * There are 5 required arguments:
 *
 * [string] the string to encode
 * [version] the version of the QR Code
 * [error correction level] an integer representing an error correction level
 * [encoding mode] an integer representing the encoding mode
 * [case sensitivity] 1 (case sensitive) or 0 (case insensitive)
 *
 * == Version
 *
 * What is the version? Each QRCode is made up of <b>modules</b> which are the
 * basic display element of a QRCode and may be made up of 1 or more pixels
 * (here, it's just 1 module is 1 pixel). Version 1 is a 21x21  module square,
 * while the maximum version 40 is 177x177 modules.  The full module reference
 * is here http://www.denso-wave.com/qrcode/vertable1-e.html
 *
 * Should you encode more text than can fit in a module, the encoder will scale
 * up to the smallest version able to contain your data. Unless you want to
 * specifically fix your barcode to a certain version, it's fine to just set
 * the version argument to 1 and let the algorithm figure out the proper size.
 *
 * == Error correction
 *
 * The following four constants can be specified for error correction levels, each
 * specified with the maximum approximate error rate they can compensate for, as
 * well as the maximum capacity of an 8-bit data QR Code with the error encoding:
 *
 * - <tt>QR_ECLEVEL_L</tt> - 7%/2953 [default]
 * - <tt>QR_ECLEVEL_M</tt> - 15%/2331
 * - <tt>QR_ECLEVEL_Q</tt> - 25%/1663
 * - <tt>QR_ECLEVEL_H</tt> - 30%/1273
 *
 * Higher error rates are suitable for applications where the QR Code is likely
 * to be smudged or damaged, but as is apparent here, they can radically reduce
 * the maximum data capacity of a QR Code.
 *
 * == Encoding mode
 *
 * There are 4 possible encodings for a QR Code which can modify the maximum
 * data capacity. These are specified with four possible Constants, each
 * listed here with the maximum capacity available for that encoding at the
 * lowest  error correction rate.
 *
 * - <tt>QR_MODE_NUM</tt> - Numeric/7089
 * - <tt>QR_MODE_AN</tt> - Alphanumeric/4296
 * - <tt>QR_MODE_8</tt> - 8-bit ASCII/2953 [default]
 * - <tt>QR_MODE_KANJI</tt>  - Kanji (JIS-1 & 2)/1817
 *
 * Note that the QR Code specification seemingly predates the rise and triumph
 * of UTF-8, and the specification makes no requirement that writers and
 * readers use ISO-8859-1 or UTF-8 or whatever to interpret the data in a
 * barcode. If you encode in UTF-8, it might be read as ISO-8859-1 or not.
 *
 * == Case sensitivity
 *
 * Encoding can either be case sensitive (1) or not (0). Without case
 * sensitivity turned on, many decoders will view all alphabetic characters as
 * uppercase.
 */
static VALUE qr_initialize(VALUE self, VALUE _string, VALUE _version, VALUE _eclevel, VALUE _hint, VALUE _casesensitive) {
  const char *string = StringValuePtr(_string);
  int version = FIX2INT(_version);
  int eclevel = FIX2INT(_eclevel);
  int hint = FIX2INT(_hint);
  int casesensitive = FIX2INT(_casesensitive);
  int error = errno;
  QRcode *code;

  QRinput *input = QRinput_new2(version, eclevel);
  QRinput_append(input, hint, strlen(string), string);
  code = QRcode_encodeInput(input);
  QRinput_free(input);

  if (errno != error && errno == 22) {
    rb_raise(rb_eArgError, "input was too long");
  }
  if (NULL == code) {
    rb_raise(rb_eArgError, "could not encode input");
  }

  if (DATA_PTR(self)) {
    QRcode_free(DATA_PTR(self));
    DATA_PTR(self) = NULL;
  }

  DATA_PTR(self) = code;
  return self;
}

/* Allocate memory */
static VALUE qr_alloc(VALUE klass) {
  QRcode *code;
  VALUE obj = Data_Make_Struct(klass, QRcode, NULL, qrcode_free, code);
  return obj;
}

void Init_qrencoder_ext()
{
    mQREncoder = rb_define_module("QREncoder");
    cQRCode = rb_define_class_under(mQREncoder, "QRCode", rb_cObject);

    rb_define_method(cQRCode, "initialize_copy", qr_init_copy, 1);
    rb_define_method(cQRCode, "initialize", qr_initialize, 5);
    rb_define_alloc_func(cQRCode, qr_alloc);

    rb_define_method(cQRCode, "width", _width, 0);
    rb_define_method(cQRCode, "version", _version, 0);
    rb_define_method(cQRCode, "data", _data, 0);
    rb_define_method(cQRCode, "pixels", _pixels, 0);
    rb_define_method(cQRCode, "points", _points, 0);

    rb_define_alias(cQRCode, "height", "width");
}
